sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,726 @@
+	// represented in the sub-buckets.
+	//
+	// To bring these two into alignment, we can roll back to version N.
+	// It might be possible to roll forward in some cases (particularly if
+	// only one item can be added at a time?) but the crash must have
+	// occurred before PutBucket() returned, so the new data need not be saved.
+	//
+	// If we load a bucket that has sub-buckets but is non-empty,
+	// delete all of the sub-buckets.  (The current implementation recurses
+	// so there may be more than one level of sub-buckets created by shardBucket.)
+
+	// Visit the buckets using inorder traversal, parents before children.
+	sort.Strings(keys)
+
+	nonemptyParent := "NOT_A_PREFIX"
+
+	s.bucketsCacheLock.Lock()
+	defer s.bucketsCacheLock.Unlock()
+	for _, key := range keys {
+		if strings.HasPrefix(key, nonemptyParent) {
+			s.Logger.Warn("detected shadowed bucket, removing", "key", key, "parent", nonemptyParent)
+			s.BucketStorageView.Delete(ctx, key)
+			continue
+		}
+
+		// Read from the underlying view
+		storageEntry, err := s.BucketStorageView.Get(ctx, key)
+		if err != nil {
+			return errwrap.Wrapf("failed to read packed storage entry: {{err}}", err)
+		}
+		if storageEntry == nil {
+			return fmt.Errorf("no data found at bucket %s", key)
+		}
+		bucket, err := s.DecodeBucket(storageEntry)
+		if err != nil {
+			return err
+		}
+		if !bucket.HasShards {
+			nonemptyParent = key
+		}
+
+		s.bucketsCache.Insert(s.GetCacheKey(bucket.Key), bucket)
+	}
+
+	s.addAllBaseBucketsToCache()
+	return nil
+}
+
+// Ensure all base buckets are present on the cache, even if they
+// are not present in storage.  This simplifies acquiring locks for
+// multiple buckets because every necessary bucket is already present
+// in bucketsCache. We don't have to worry about adding them on-demand
+// in the middle of that process, and if any object is added to these
+// placeholders they'll get saved to stoarge.
+func (s *StoragePackerV2) addAllBaseBucketsToCache() {
+	for _, bucketKey := range s.getAllBaseBucketKeys() {
+		if _, present := s.bucketsCache.Get(bucketKey); !present {
+			lock := locksutil.LockForKey(s.storageLocks, bucketKey)
+			s.bucketsCache.Insert(bucketKey,
+				&LockedBucket{
+					LockEntry: lock,
+					Bucket: &Bucket{
+						Key:       bucketKey,
+						ItemMap:   make(map[string][]byte),
+						HasShards: false,
+					},
+				})
+		}
+	}
+}
+
+func (s *StoragePackerV2) BucketKeys(ctx context.Context) ([]string, error) {
+	var retErr error
+	s.prewarmCache.Do(func() {
+		diskBuckets, err := logical.CollectKeys(ctx, s.BucketStorageView)
+		if err != nil {
+			retErr = err
+			return
+		}
+		retErr = s.preloadBucketsFromDisk(ctx, diskBuckets)
+	})
+	if retErr != nil {
+		return nil, retErr
+	}
+
+	ret := make([]string, 0, 256)
+	s.bucketsCacheLock.RLock()
+	s.bucketsCache.Walk(func(s string, _ interface{}) bool {
+		ret = append(ret, s)
+		return false
+	})
+	s.bucketsCacheLock.RUnlock()
```

Comment:
I think this `RUnlock` should be inside Do's `func`.

Reply:
Why?

preloadBucketsFromDisk does its own (write-)locking on bucketsCacheLock, and bucketsCacheLock is not acquired in the anonymous function itself.

Reply:
> bucketsCacheLock is not acquired in the anonymous function itself.

You are calling `s.bucketsCacheLock.RLock()` inside the lambda func, in [L175](https://github.com/hashicorp/vault/pull/6998/files#diff-d460b49b4e599a77c84c96a74553b2d8R175).

Reply:
Oh wait nvm...